MINOR: upgrade to jdk8 8u202

jarekr opened the pull request:
Upgrade from 171 to 202. Unpack and install directly from a cached tgz rather than going via the installer deb from webupd8. The installer is still on 8u919 while we want 202.

Testing via kafka branch builder job

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/2305/

Commit:
MINOR: jdk8 ppa updated to use 8u191

Commit:
rename dir so for loop works

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,41 @@ set -ex
-    apt-get -y update
+path_to_jdk_cache() {
+  jdk_version=$1
+  echo "/var/cache/oracle-jdk8-installer/jdk-${jdk_version}-linux-64.tar.gz"
```
we should probably just remove the whole caching piece.

the caching only makes sense in the context of the webupd8team repository and local use of vagrant for system tests. this patch removes use of the old webupd8team installer (which I think is the right long term move given pains we've seen over time). the cache also mostly only made sense when running vagrant locally against virtualbox, where the cached data could be shared against, e.g., 10 VMs by mounting a local directory into each of the VMs, thus avoiding re-downloading the data. I think people have mostly moved to either using local docker builds/tests *or* nightlies against EC2 that don't benefit from the caching and where we should only be building once for an AMI and then launching that N times.

jarekr replied:
I removed the caching piece

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,41 @@ set -ex
-    apt-get -y update
+path_to_jdk_cache() {
+  jdk_version=$1
+  echo "/var/cache/oracle-jdk8-installer/jdk-${jdk_version}-linux-64.tar.gz"
```
oh, and i forgot to mention what i *really* noticed originally, which is that this path has a fixed `oracle-jdk8-installer` in it, which doesn't match the `jdk_version` variable.

also, the path was originally meant to match https://github.com/apache/kafka/blob/5040ec02ddb3f14aa02d66402c2408a21b77a257/Vagrantfile#L85, which if we're moving away from due to moving away from the webupd8team installler, we can also clean up.

we could use a fixed `/tmp` directory here if we wanted to now since not using the installer means it doesn't need a fixed path and the jdk download filenames should all be unique

jarekr replied:
I just moved the download to /tmp

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,41 @@ set -ex
-    fi
+if [ -z `which javac` ]; then
+    apt-get -y update
+    apt-get install -y software-properties-common python-software-properties binutils gsfonts gsfonts-x11 java-common libfontenc1 libxfont1 x11-common xfonts-encodings xfonts-utils
```
how did we come up with this list of packages? anything `x11` seems suspicious -- do we really need all of that since everything we run is headless? pulling in `x11` tends to result in bloated images.

jarekr replied:
I had gotten the list from a job that was still installing via the webupd8 deb and copied the list of packages to be installed minus the installer packages themselves, not wanting to make too many changes at once (e.g. keep installing the same dependencies, just replace the installer package with our own install). I changed this to just java-common as I think that is probably something we need.

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,41 @@ set -ex
+    mkdir -p /opt/jdk
+    cd /opt/jdk
+    rm -rf 8
+    jdk_version='8u202'
```
can we move this (and the jdk versions below) somewhere generic and overridable, e.g. passed in from the Vagrantfile via args or env?

we'll want to be able to override this to jdk11 or other versions, so we'll need a mechanism to adjust these

jarekr replied:
Yep, it's now overridable in Vagrantfile / Vagrantfile.local

avocader commented on vagrant/base.sh:
```diff
@@ -20,38 +20,41 @@ set -ex
+      name=$(basename $bin)
+      update-alternatives --install /usr/bin/$name $name $bin 1091 && update-alternatives --set $name $bin
+    done
+    echo "export JAVA_HOME=/opt/jdk/8\nexport PATH=\$PATH:\$JAVA_HOME/bin" > /etc/profile.d/jdk.sh
```
Should it be `echo -e "export...`?

jarekr replied:
yes, thanks!

Commit:
address PR comments

Allow overriding (and support jdk 10+) from commandline using JDK_MAJOR and
JDK_FULL env vars.

Commit:
make overridable in / draw default from Vagrant file

Commit:
syntax

Commit:
use path var to reduce clutter

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,43 @@ set -ex
+    cd /opt/jdk
+    rm -rf $JDK_MAJOR
+    fetch_jdk_tgz $JDK_FULL
+    tar x --strip-components=1 zf $(path_to_jdk_cache $JDK_FULL)
```
this fails for me in testing, not sure if it has to do with tar implementations varying but i think only the first parameter can be a function without the `-` prefix. the `zf` part needs `-`s.

ewencp commented on vagrant/base.sh:
```diff
@@ -20,38 +20,43 @@ set -ex
+    echo "===> Installing JDK..." 
+
+    mkdir -p /opt/jdk
+    cd /opt/jdk
```
i think subsequent `tar` command is leaving the entire jdk in `/opt/jdk` without a jdk version directory. this line and the line above should include `$JDK_MAJOR` as a directory.

ewencp reviewed (commented):
@jarekr this is close, with two fixes i mentioned a simple local test against virtualbox seems to leave a clean java setup.

Commit:
fix tar command, unpack jdk tar in JDK_MAJOR dir

ewencp committed:
Fix typo

ewencp reviewed (approved):
@jarekr I think your previous system test job run is taking awhile but am surprised it is working since I found and fixed a typo.

I just tested against a Virtualbox Vagrant (though that admittedly doesn't cover the AMI change) and with the typo fixed, things seem ok. Do we need a follow up test or are we ok to merge this (presumably things are just broken now anyway, so not sure we can make it any worse...).

Do we know how far back we need this? It looks like the jdk8 install was added on the 1.0 branch even though we didn't officially support it until 2.0.

jarekr commented:
ooh, thanks.

I think we can merge this. Someone (pretty sure it wasn't me) merged my proposed update to jdk 191, which should have fixed the install (for now).

I was going to say we need it back to and including 1.0, seems easy enough to cherry pick?

ewencp commented:
1.0 is fine, just need to know target. with cherry-picking, if nobody has validated backporting, it can be messier and require more steps. if it cherry-picks cleanly, then it's easy. i'll cherry-pick back as long as it works back to 1.0, if anything fails, we'll follow up.
